### ISSUE TITLE
Decoded value differs from input

### ISSUE THREAD

**Issue by s0meone** [mtth/avsc#374]
Hi there!

While changing our data serialisation from JSON to Avro an issue raised in our application where we accidentally try to  store invalid UTF8 strings. We didn't notice this issue before because JSON serialises and deserialises the invalid strings just fine.
Avsc handles these strings a bit different which causes the decoded value to be different than the original input.

I've prepared an example demonstrating this behaviour below. The input `string` is a valid UTF8 string which we then accidentally split by byte instead of character. These individual bytes are non valid UTF8 strings but can be parsed correctly by JSON.

```js
const avroType = avro.Type.forSchema({
  name: "charArray",
  type: "array",
  items: "string",
});

const string = "𝐼𝑁𝐷𝑈𝑆𝑇𝑅𝑌";
const stringSplit = string.split("");

const endodedJson = JSON.stringify(stringSplit);
const decodedJson = JSON.parse(endodedJson);

const encodedAvro = avroType.toBuffer(stringSplit);
const decodedAvro = avroType.fromBuffer(encodedAvro);

console.log({
  string,
  stringSplit,
  stringSplitJoin: stringSplit.join(""),
  decodedJson,
  decodedJsonJoin: decodedJson.join(""),
  decodedAvro,
  decodedAvroJoin: decodedAvro.join(""),
});
```

The output of above:

```
{
  string: '𝐼𝑁𝐷𝑈𝑆𝑇𝑅𝑌',
  stringSplit: [
    '�', '�', '�', '�',
    '�', '�', '�', '�',
    '�', '�', '�', '�',
    '�', '�', '�', '�'
  ],
  stringSplitJoin: '𝐼𝑁𝐷𝑈𝑆𝑇𝑅𝑌',
  decodedJson: [
    '�', '�', '�', '�',
    '�', '�', '�', '�',
    '�', '�', '�', '�',
    '�', '�', '�', '�'
  ],
  decodedJsonJoin: '𝐼𝑁𝐷𝑈𝑆𝑇𝑅𝑌',
  decodedAvro: [
    '���', '���', '���',
    '���', '���', '���',
    '���', '���', '���',
    '���', '���', '���',
    '���', '���', '���',
    '���'
  ],
  decodedAvroJoin: '������������������������������������������������'
}
```

I did not expect this behaviour from Avsc and expected the output to be exactly the same as the input. Is my assumption wrong or is this something this library (or documentation) could improve?

**Comment by mtth** [comment]
Hey @s0meone. Great observation. It's actually slightly different: `string.split` here doesn't split by byte but by UTF-16 codeunit, which breaks surrogate pairs (see [this StackOverflow question](https://stackoverflow.com/a/34717402) for example). The code above therefore produces invalid (UTF-16) strings and behavior when encoding these is undefined.

The built-in JSON encoder handles these invalid strings by escaping them (e.g. `"\\ud835"`). This is lossless and allows the individual parts to be recombined and produce the original input. Unfortunately `Buffer.write` methods, which `avsc` uses under the hood, don't preserve broken surrogate pairs when encoding with UTF-8: their encoding gets replaced by invalid markers. `avsc` inherits this limitation as your example demonstrates.

This is obviously not great but I'm afraid there isn't much we can do without paying a performance penalty for validating each string beforehand. This is unlikely to be acceptable on the encoding path but I can consider adding an option to `isValid` which performs this check if you think it'd be useful.

**Comment by s0meone** [comment]
Thank you @mtth for your helpful answer. It made me realise I actually know less about JavaScript Unicode strings than I thought. So I've looked up some more [info](https://dmitripavlutin.com/what-every-javascript-developer-should-know-about-unicode/) and found this definition:

> Probably the most important concept about Unicode in JavaScript is to treat strings as sequences of code units, as they really are. 

After some more reading I also found your statement about the invalid markers in the [documentation](https://nodejs.org/api/buffer.html#buftostringencoding-start-end) of `Buffer`, which describes the problem exactly:

> If encoding is 'utf8' and a byte sequence in the input is not valid UTF-8, then each invalid byte is replaced with the replacement character U+FFFD.

I understand that validating each string will drop performance and I don't think it will be necessary to validate this every time as it's basically normal NodeJS behaviour. But it might be helpful to add this behaviour to the documentation of `avsc` somewhere?

Thanks again for your help!
